### PR TITLE
[ BB2-604 ] Removing implicit flow docs. Adding PKCE docs

### PR DESCRIPTION
**Pull request by nbragdon**
**JIRA Ticket:**
https://jira.cms.gov/browse/BB2-604

**User Story or Bug Summary:**
Better documentation on how to implement the PKCE Extension so that I can implement the promoted feature

NOTE: 6/10/2021 The changes from the following PR has been merged in to this one and has been closed:  https://github.com/CMSgov/bluebutton-site-static/pull/240#

### What Does This PR Do?

1. Removes documentation to implicit flow since we no longer support it
2. Removes the "Mobile Support PKCE" section, we don't support a non-backend flow and that was only creating more confusion for users
3. Add in a new PKCE section with a code example, along with request templates
4. Added in a message in the normal OAuth flow section saying that using PKCE is now recommended
5. Added PKCE to the postman example

### What Should Reviewers Watch For?

- Make sure there are no issues removing the sections I did
- Make sure there are no typos/it reads well. Especially in the new PKCE section.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls?  **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **Yes** or **No**.

### What Needs to Be Merged and Deployed Before this PR?

No prereqs

### Submitter Checklist

I have gone through and verified that...:

* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by micahtaylor — COMMENTED** [review]
Seems good from my perspective but I look to @dtisza1 or @JFU-GIT for content.

**Comment by njdister** [comment]
Logically it's looking good, I'm just not sure if external devs would be confused by the typescript example as the rest of the examples are curl or pseudocode.

**Comment by nbragdon** [comment]
@njdister Do you think we shouldn't do any code samples? Or basically just wait to do them until we can do them across the board (rather than piecemeal)?

**Comment by njdister** [comment]
> @njdister Do you think we shouldn't do any code samples? Or basically just wait to do them until we can do them across the board (rather than piecemeal)?

Yeah I'd probably defer code samples until we can update them across the board to make the documentation consistent.  Any opinion @dtisza1 @JFU-GIT ?

**Comment by dtisza1** [comment]
> > @njdister Do you think we shouldn't do any code samples? Or basically just wait to do them until we can do them across the board (rather than piecemeal)?
> 
> Yeah I'd probably defer code samples until we can update them across the board to make the documentation consistent. Any opinion @dtisza1 @JFU-GIT ?

@njdister @nbragdon I agree here, with examples being in CURL or pseudo code.  

Related to the PKCE instruction, I didn't see "code_challenge_method=S256" referenced.  In the current docs (removed section), it looks like this has a typo as "codechallengemethod".

I was also wondering if we should remove the link to the Android sample app or update the description for it?
At the time, it was created more to demonstrate & test PKCE and not as a secure implementation example.  It might need a warning something like:  "This sample application was created to demonstrate PKCE and custom URI schemes; however, it is up to your implementation to secure the client_secret in the auth flow."  

Otherwise, the changes are looking good to me. We should have a team review discussion on a next stand up too.

**Comment by njdister** [comment]
> > > @njdister Do you think we shouldn't do any code samples? Or basically just wait to do them until we can do them across the board (rather than piecemeal)?
> > 
> > 
> > Yeah I'd probably defer code samples until we can update them across the board to make the documentation consistent. Any opinion @dtisza1 @JFU-GIT ?
> 
> @njdister @nbragdon I agree here, with examples being in CURL or pseudo code.
> 
> Related to the PKCE instruction, I didn't see "code_challenge_method=S256" referenced. In the current docs (removed section), it looks like this has a typo as "codechallengemethod".
> 
> I was also wondering if we should remove the link to the Android sample app or update the description for it?
> At the time, it was created more to demonstrate & test PKCE and not as a secure implementation example. It might need a warning something like: "This sample application was created to demonstrate PKCE and custom URI schemes; however, it is up to your implementation to secure the client_secret in the auth flow."
> 
> Otherwise, the changes are looking good to me. We should have a team review discussion on a next stand up too.

Yeah I wouldn't mind removing references to the android sample app since it has now been archived.

**Comment by nbragdon** [comment]
@njdister @njdister Made the updates suggested. 

Also just my take on the Typescript example: We can't really do a curl example for PKCE, since its not an API request. The pseudocode formula was there, but I know the first thing I did when reading it was to go and try to find a real code example via google. It might be a good place to add in our first code example. Could also do multiple languages if you think that would be better.

**Comment by micahtaylor** [comment]
Of note on this from user feedback session, end users (developers) have asked for code examples. I believe they will not be confused even if it isn't in their preferred language. 
I also believe that incorporating code examples is the direction we are heading with the updated documentation so this is a low risk way for us to begin that journey.

**Comment by njdister** [comment]
> Of note on this from user feedback session, end users (developers) have asked for code examples. I believe they will not be confused even if it isn't in their preferred language.
> I also believe that incorporating code examples is the direction we are heading with the updated documentation so this is a low risk way for us to begin that journey.

@micahtaylor Well I stand corrected!  Thanks for this insightful feedback.

@nbragdon Maybe there is a way to leave the example in, but add a footnote that this is part of a larger effort to revamp the code examples?

**Comment by JFU-GIT** [comment]
@team, the doc changes look good, originally I was thinking delegating the sample code to external resources e.g. auth0 to avoid inventing wheels BUT since user feedback asked about that, then should go for it, understand that 'one stop' experience helps when starting something.

Can even push further by extend the testclient to provide pkce option and make source code (in python) available - a zip that can download and play with... just some 2c, have a good Friday

**Review by njdister — APPROVED** [review]

File: _pages/developers.md [hunk 1]
```diff
@@ -283,8 +261,6 @@ For example:
 tld.app.subdomain[.subsubdomain]:/callback/path/endpoint
 ```
 
-A coding example of an OAuth 2.0 and PKCE flow is available here: [Authorization Code with PKCE Flow - OAuth 2.0 Playground](https://www.oauth.com/playground/authorization-code-with-pkce.html)
-
 The Blue Button 2.0 API engineering team has also created a sample Android application. You can review or fork the code here: [https://github.com/CMSgov/bluebutton-sample-client-android](https://github.com/CMSgov/bluebutton-sample-client-android)
```
Comment (nbragdon) [inline _pages/developers.md:264]:
```suggestion
```

File: _pages/developers.md [hunk 2]
```diff
@@ -404,50 +381,84 @@ Response
...
+* Your client must validate that the "state" parameter equals the one that was previously stored (from your initial authorize request). If you receive a response that does not match, your client should exit the flow (since there may be a compromise). 
+* Your client has now successfully correlated the original request with the call-back response received from the authentication.
 
+If implementing PKCE, clients may opt to not use state, since similar CSRF protection is provided by it. However, the state may be useful to your application for non-security type purposes, such as encoding other information in it for the authorization response.
+
+The following article has additional information about usage and the type of attack it helps to mitigate: https://auth0.com/docs/protocols/state-parameters
+
+### Adding PKCE
+
+PKCE stands for Proof Key for Code Exchange. PKCE is an extension of the Authorization code flow that protects from authorization code injection attacks. Implementing PKCE involves creating a secret that is used when exchanging the authorization code to obtain an access token. You can read more about it from the [OAuth docs](https://oauth.net/2/pkce/)
```
Comment (nbragdon) [inline _pages/developers.md:404]:
```suggestion
PKCE stands for Proof Key for Code Exchange. PKCE is an extension of the Authorization code flow that protects from authorization code injection attacks. Implementing PKCE involves creating a secret that is used when exchanging the authorization code to obtain an access token. You can read more about it from the [OAuth docs](https://oauth.net/2/pkce/).
```

**Review by dtisza1 — APPROVED** [review]
This looks good to me with Nick's change suggestions merged in.

Had some optional suggestions related to period usage in our slack thread, if others think those are good to do.

**Pull request merged**